# Labelling

In [ ]:
import deeplabcut
from PathGeneral_Func import GetFilePaths
import os

Loading DLC 2.2.2...


### Gather videos to analyse

In [ ]:
#Videos_Dir      = r"O:\OneDrive - MRC Laboratory of Molecular Biology\Resources\Workshops\DLC_20220916\Data\Videos\Avis"
Videos_Dir      = r"C:\Users\Edwar\Documents\deeplabcut\Motor videos for training\DLC Motor Training Set\\"

Videos_Paths    = GetFilePaths(Videos_Dir, '.avi')
#Videos_Paths = [r"D:\MTCE1.3e\20210621\Concatenated\Movies\Cameras\Eye\Eye_Cropped.avi"] # If only 1 movie

print(Videos_Paths)

### Create Project

In [5]:
Project_Name = 'DLC_TRN_191022_v2'
Experimenter = "ECH"

#Project_Dir = r"O:\OneDrive - MRC Laboratory of Molecular Biology\Resources\Workshops\DLC_20220916\Data"
Project_Dir = r"C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022"

Config_Path = deeplabcut.create_new_project(Project_Name, Experimenter, 
                                            Videos_Paths, videotype = '.avi', copy_videos = True,
                                            working_directory = Project_Dir)

Created "C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\videos"
Created "C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\labeled-data"
Created "C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\training-datasets"
Created "C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\dlc-models"
Copying the videos
C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\videos\010522_GA_NBH_M00708462_1R_CD_18wk.avi
C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\videos\010522_GA_RML_M00708451_NM_CD_18wk.avi
C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\videos\010522_GA_RML_M00708453_1R_CD_18wk.avi
C:\Users\Edwar\Documents\deeplabcut\DLC_Training_191022\DLC_TRN_191022_v2-ECH-2022-10-22\videos\010522_GA_RML_M00708479_1L_CD_SLN_1

### Add manually body parts in the config.yaml file (Change also move2corner to false)

In [ ]:
# Open config file with notepad

In [ ]:
# enter body parts to label

In [ ]:
# Adjust numFrames 2 pick

In [ ]:
# Change move to corner to false

### Extract/Select which frames to manually annotate

In [6]:
#Config_Path = r"D:\MTCM2.1b\20220531\Eye-DdM-2022-06-23\config.yaml"
cluster_step = 10 # consider only every 10th frame for the kmeans
deeplabcut.extract_frames(Config_Path, algo='kmeans', userfeedback=False, cluster_step=cluster_step , cluster_resizewidth=30, slider_width=25)


Config file read successfully.
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:11, 35.38it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.27  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:12, 30.89it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:13, 29.80it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:11, 33.78it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:12, 32.17it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:11, 34.31it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:12, 30.39it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:14, 26.63it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.17  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:37, 10.35it/s]


Kmeans clustering ... (this might take a while)
Frame 3470  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.17  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:36, 10.67it/s]


Kmeans clustering ... (this might take a while)
Frame 3520  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 24.48it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:16, 24.19it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:18, 21.41it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:18, 21.32it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 21.95it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 22.05it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:18, 21.14it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 22.07it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 21.77it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 21.78it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 131.37  seconds.
Extracting and downsampling... 395  frames from the video.


395it [00:15, 24.81it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:18, 20.94it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 24.70it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:38, 10.13it/s]


Kmeans clustering ... (this might take a while)
Frame 3820  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:38, 10.17it/s]


Kmeans clustering ... (this might take a while)
Frame 3500  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:40,  9.73it/s]


Kmeans clustering ... (this might take a while)
Frame 3710  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:38, 10.15it/s]


Kmeans clustering ... (this might take a while)
Frame 3450  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.17  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:38, 10.17it/s]


Kmeans clustering ... (this might take a while)
Frame 3520  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:16, 24.24it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 24.82it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 24.67it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.13  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 21.99it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:18, 21.54it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 25.01it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:17, 22.59it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.07  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:16, 24.25it/s]


Kmeans clustering ... (this might take a while)
Extracting frames based on kmeans ...
Kmeans-quantization based extracting of frames from 0.0  seconds to 130.1  seconds.
Extracting and downsampling... 391  frames from the video.


391it [00:15, 24.58it/s]


Kmeans clustering ... (this might take a while)
Frame 3900  not found!
Frames were successfully extracted, for the videos listed in the config.yaml file.

You can now label the frames using the function 'label_frames' (Note, you should label frames extracted from diverse videos (and many videos; we do not recommend training on single videos!)).


## Label frames

In [ ]:
deeplabcut.label_frames(Config_Path)

# Transfer to big computer

### Transfer Data to beegfs3

In [ ]:
#mount beegfs3 smb://beegfs3/beegfs3 or \\beegfs3\beegfs3 and create user folder

In [ ]:
# Or transfer via atg_import

In [ ]:
# Move Project file to Beegfs

### log in to hex

In [ ]:
# Via atg or directly

In [ ]:
# Using terminal or x2go

### install anaconda and DLC

In [ ]:
# dowload anaconda
# bash anconcondaFile.sh

In [ ]:
# dowload deeplab cut yaml file
# remove the [gui] part from inside the file
# Install with: conda env create -f DEEPLABCUT.yaml

### Set up build ( cuda 10.1, tensorflow 2.3 cudnn=7.6 and python=3.8 )

In [ ]:
# Activate environment
# Install cudnn:
# conda install cudnn==7.6 

In [ ]:
# Install tensorflow
# pip install tensorflow==2.3

### Prepare python and bash file for server

In [ ]:
# Copy DLCRoutine.py and BashScript.sh to Hex desktop

In [ ]:
# Change project path in config yml to new path : /beegfs3// + name

In [ ]:
# Upate config variable in DLCroutine.py file
# ex: config = "/beegfs3/dmalmazet/Eye_Cropped/Eye-DdM-2021-12-10/config.yaml"

In [ ]:
# edit BashScript.sh if want to change cluster settings such as number of gpu to use

### Run sbatch script

In [ ]:
# On Hex
# open comand prompt
# cd to desktop
# Run sbatch BashScript.sh

In [ ]:
# check cluster usage: http://nagios2/qinfo/

In [ ]:
# Rules
# Eight GPUs in total per user (i.e. 1 full Asus GPU node, or spread thinly over several GPU nodes).

# Back to labelling computer

In [ ]:
# Check results with labeled movies

### Load Results

In [ ]:
import pandas as pd
from PathGeneral_Func import GetFilePaths

#% Load deeplabcut data
DLC_Results_Dir     = r"O:\OneDrive - MRC Laboratory of Molecular Biology\Resources\Workshops\DLC_20220916\Data\Results"
DLC_Results_Paths   = GetFilePaths(DLC_Results_Dir, 'h5')

Video_Num = 0
df = pd.read_hdf(DLC_Results_Paths[Video_Num])
print("\nLoading DLC file:", DLC_Results_Paths[Video_Num])
print(df.dtypes)

### Convert data into numpy array

In [ ]:
import numpy as np

# Adjust to correct position
Pos_Temporal    = df.to_numpy()[:, 0:3] # _Col_Row_Likelihood
Pos_Bottom      = df.to_numpy()[:, 3:6]
Pos_Nasal       = df.to_numpy()[:, 6:9]
Pos_Top         = df.to_numpy()[:, 9:13]

print(Pos_Temporal.shape)

### Filter data

In [ ]:
from DLC_Functions import MedianFilterPos

# Median filter
Pos_Temporal[:, :2] = MedianFilterPos(Pos_Temporal[:, :2])
Pos_Nasal[:, :2]    = MedianFilterPos(Pos_Nasal[:, :2])
Pos_Bottom[:, :2]   = MedianFilterPos(Pos_Bottom[:, :2])
Pos_Top[:, :2]      = MedianFilterPos(Pos_Top[:, :2])

### Compute Pupil centre position and diameter

In [ ]:
#% Pupil center
nFrames = Pos_Temporal.shape[0]
Pupil_centre = np.empty((nFrames, 2), dtype = np.float64)

Pupil_centre[:, 0] = np.mean( np.concatenate( (Pos_Temporal[:, 0:1], Pos_Nasal[:, 0:1], Pos_Top[:, 0:1], Pos_Bottom[:, 0:1]), axis=1 ), axis=1 ) 

Pupil_centre[:, 1] = np.mean( np.concatenate( (Pos_Temporal[:, 1:2], Pos_Nasal[:, 1:2], Pos_Top[:, 1:2], Pos_Bottom[:, 1:2]), axis=1) , axis = 1)

In [ ]:
#% Pupil diameter

#% transform to complex
Pos_Temporal_Complex         = np.empty(nFrames, dtype=complex)
Pos_Temporal_Complex.real    = Pos_Temporal[:,0]
Pos_Temporal_Complex.imag    = Pos_Temporal[:,1]

Pos_Nasal_Complex         = np.empty(nFrames, dtype=complex)
Pos_Nasal_Complex.real    = Pos_Nasal[:,0]
Pos_Nasal_Complex.imag    = Pos_Nasal[:,1]

Pos_Top_Complex         = np.empty(nFrames, dtype=complex)
Pos_Top_Complex.real    = Pos_Top[:,0]
Pos_Top_Complex.imag    = Pos_Top[:,1]

Pos_Bottom_Complex         = np.empty(nFrames, dtype=complex)
Pos_Bottom_Complex.real    = Pos_Bottom[:,0]
Pos_Bottom_Complex.imag    = Pos_Bottom[:,1]

#% Pupil diameter
Pupil_Diameter = ( np.abs(Pos_Temporal_Complex - Pos_Nasal_Complex) + np.abs(Pos_Top_Complex - Pos_Bottom_Complex) )/2

### Plot pupil diameter and pupil centre position

In [ ]:
import matplotlib.pyplot as plt
%matplotlib qt

In [ ]:
# Plot pupil diameter and pupil centre position
#%  Plot
fig, ax = plt.subplots(nrows=3, ncols=1, sharex=True)

ax[0].plot(Pupil_Diameter)
ax[0].set_title('Pupil diameter')

ax[1].plot(Pupil_centre[:, 0])
ax[1].set_title('Col')

ax[2].plot(Pupil_centre[:, 1])
ax[2].set_title('row')

# Extra

In [ ]:
# Refine
deeplabcut.extract_outlier_frames(Config_Path , Videos , outlieralgorithm='manual', comparisonbodyparts='Cricket')

deeplabcut.refine_labels(Config_Path)

deeplabcut.merge_datasets(Config_Path)